### PR TITLE
Clean up repo

### PR DESCRIPTION
**Pull request by PatrickRMiles**
This branch removes deprecated execution paths and tightens several pieces of benchmark plumbing before publication. The changes reduce unsupported configuration surface area, make logging and failure behavior more consistent across ranks, and remove special cases that no longer represent supported benchmark modes.

### Distributed execution is now required

- Removed the user-facing `dist` option from the default config and README examples.
- Updated benchmark, worker, and trainer paths to assume distributed execution.
- Kept one-rank runs supported as distributed singleton jobs rather than a separate non-distributed mode.
- Added a clear config error if an old config explicitly sets `dist: 0`.
- Simplified trainer sampler setup, warmup barriers, checkpoint coordination, validation reduction, and DistConv/DDP setup now that those paths no longer branch on `config.dist`.
- Improved the DistConv world-size check by raising a `ValueError` when `world_size` is not divisible by the total DistConv shard count.

### Deprecated zero-fractal-class behavior was removed

- `n_categories` must now be a positive integer at config load and CLI override time.
- Removed the `BCEWithLogitsLoss` fallback used only by the deprecated zero-category case.
- Cross entropy is now always used for the benchmark segmentation task.
- Foreground Dice calculations now consistently exclude the background channel with `[:, 1:]`.
- Volume generation now sizes fractal colors from `n_categories` directly instead of guarding for the old zero-category case.

### CLI and config cleanup

- Added a benchmark CLI override for `--dc-shard-dims`, matching the existing `dc_shard_dims` config option.
- Cleaned up DistConv option names and help text so config, CLI, and runtime logging use the same terminology.
- Made `generate_fractals --datagen-batch-size` drive category search batch size instead of relying on a hard-coded value.
- Removed the stale `ScaFFold/configs/benchmark_testing.yml` config.
- Removed unused benchmark sweep helper functions that were no longer called.
- Fixed small documentation and help-text issues in the README and CLI.

### Logging and diagnostics are more consistent

- Standardized many benchmark, CLI, data-generation, trainer, evaluation, worker, and checkpoint messages on `setup_mpi_logger`.
- Replaced high-traffic `print` statements with rank-aware logger calls.
- Made MPI logger setup idempotent so repeated setup calls do not attach duplicate handlers.
- Kept logger imports lighter by moving heavy imports such as Torch, NumPy, Matplotlib, and Torch distributed into the functions that actually need them.
- Converted routine multi-rank diagnostic noise to debug logging while keeping important lifecycle messages at info or warning level.

### Dataset generation and error handling are more robust

- Fixed the dataset-generation failure path so all ranks participate in the error gather when volume generation fails. This avoids one rank raising while others hang at MPI collectives.
- Missing fractal point-cloud files now raise `FileNotFoundError` with an actionable message instead of printing and calling `sys.exit(1)`.
- Dataset reuse messages now go through the MPI logger.
- Git commit lookup failures during dataset metadata creation now emit logger warnings rather than raw prints.

### Runtime and packaging polish

- DistConvDDP device IDs now come from the selected CUDA device index, which is more robust when each rank sees a masked single-GPU view.
- The benchmark worker now destroys the process group only if it was initialized.
- `psutil>=5.9.0` is now declared in both `pyproject.toml` and `requirements.txt` because ScaFFold imports it directly.
- README examples now describe distributed-only execution, positive `n_categories`, per-rank batch size, DistConv shard settings, warmup batches, and convergence threshold.

## Behavior Changes

- Old configs with `dist: 1` continue to work because the config loader ignores the deprecated key when it is truthy.
- Old configs with `dist: 0` now fail fast with a clear message.
- Running with one process is still supported, but it should be launched as a one-rank distributed job.
- `n_categories: 0` is no longer accepted.
- Category search now honors the configured or CLI-provided `datagen_batch_size`.
- Logs should be less noisy and more consistently rank-qualified, especially in multi-rank jobs.

### CODE REVIEW

File: ScaFFold/datagen/get_dataset.py [hunk 1]
```diff
@@ -87,14 +88,18 @@ def _git_commit_short() -> str:
             .strip()
         )
     except subprocess.CalledProcessError:
-        print(
-            "Tried to get git commit id in non-git repo. No commit id will be enforced for dataset reuse."
-        )
+        if log is not None:
```
Comment (michaelmckinsey1) [inline ScaFFold/datagen/get_dataset.py]: If we are passing log at line 134, do we need to check if it is none? I don't think so.

File: ScaFFold/datagen/instance.py [hunk 2]
```diff
@@ -184,8 +187,10 @@ def main(config: Config):
     end_time = time.time()
     total_time = end_time - start_time
     if rank == 0:
-        print(
-            f"Generated {len(instances_to_generate)} instances in {total_time:.2f} seconds"
+        log.info(
+            "Generated %s instances in %.2f seconds",
+            len(instances_to_generate),
+            total_time,
         )
```
Comment (michaelmckinsey1) [inline ScaFFold/datagen/instance.py]: remove the `if rank == 0` check before logging, since this is only logging?

File: ScaFFold/utils/checkpointing.py [hunk 3]
```diff
@@ -309,8 +313,15 @@ def _write_to_disk(state_dict, last_path, best_path, is_best):
                 try:
                     torch.save(state_dict, best_path)
                 except Exception as e:
-                    print("Saving checkpoint failed. Continuing...")
-                    print(e)
+                    CheckpointManager._log_save_failure(log, e)
+
+    @staticmethod
+    def _log_save_failure(log, exc):
+        if log is not None:
```
Comment (michaelmckinsey1) [inline ScaFFold/utils/checkpointing.py]: Again, don't we always want the logging? So this will always be true?

File: ScaFFold/utils/distributed.py [hunk 4]
```diff
@@ -12,6 +12,8 @@
 #
 # SPDX-License-Identifier: (Apache-2.0)
 
+from __future__ import annotations
```
Comment (michaelmckinsey1) [inline ScaFFold/utils/distributed.py]: I don't see any forward references in this file so this can be removed
```suggestion
```

File: ScaFFold/utils/trainer.py [hunk 5]
```diff
@@ -58,9 +58,9 @@ def __init__(self, model, config, device, log):
         self.amp_device_type = self.device.type if self.device.type != "mps" else "cpu"
         self.amp_dtype = AMP_DTYPE
         self.use_grad_scaler = False
-        self.world_size = get_world_size(required=self.config.dist)
-        self.world_rank = get_world_rank(required=self.config.dist)
-        self.local_rank = get_local_rank(required=self.config.dist)
+        self.world_size = get_world_size(required=True)
```
Comment (michaelmckinsey1) [inline ScaFFold/utils/trainer.py:61]: Are there any cases anymore where we call this with `required=False`? maybe we just make the function definition default to True instead

File: ScaFFold/utils/trainer.py [hunk 6]
```diff
@@ -259,9 +249,7 @@ def _autocast_kwargs(self, enabled=None):
     @staticmethod
     def _foreground_dice_mean(dice_scores):
         """Match optimization to the reported validation metric by excluding background."""
-        if dice_scores.size(1) > 1:
-            return dice_scores[:, 1:].mean()
-        return dice_scores.mean()
+        return dice_scores[:, 1:].mean()
```
Comment (michaelmckinsey1) [inline ScaFFold/utils/trainer.py:253]: hmm I don't know why this was checking `if dice_scores.size(1) > 1:`, but with >1 classes always this should be fine as it is now.

**Review by michaelmckinsey1 — CHANGES_REQUESTED** [review]
In general I'm curious about your thoughts on the checks for `log is not None`. Are there cases where we plan to call these functions without the logger enabled?

Otherwise I think we just omit the check everywhere.

File: ScaFFold/utils/trainer.py (hunk 5 above)
Comment (PatrickRMiles) [reply ScaFFold/utils/trainer.py:61]: Since there are a lot of other pieces of the codebase that use get_world_size, I hesitate to change the default behavior (at least in this PR). I'd prefer to leave it as is

File: ScaFFold/utils/distributed.py [hunk 7]
```diff
@@ -183,9 +174,12 @@ def initialize_dist(
     else:
         raise ValueError(f'Unrecognized scheme "{rendezvous}"')
 
-    print(
-        f"distributed.py: rank {get_world_rank()} / {get_world_size()} calling init_process_group()"
-    )
+    if log is not None:
```
Comment (michaelmckinsey1) [inline ScaFFold/utils/distributed.py]: one more then I am good to approve

**Review by michaelmckinsey1 — APPROVED** [review]
